File: tools/palconv.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char ** argv) {
	if (argc != 3) {
		fputs("Expected two arguments", stderr);
		exit(1);
	}

	FILE * outfile = fopen(argv[1], "wb");
	FILE * infile = fopen(argv[2], "rb");
	if (!infile) {
		perror("input file");
		exit(1);
	}
	if (!outfile) {
		perror("output file");
		exit(1);
	}

	while (1) {
		uint16_t color = (uint8_t) fgetc(infile);
		color |= (uint8_t) fgetc(infile) << 8;
		if (feof(infile)) break;
		fputc((color >> 5 & 0b11111) * 8, outfile);
		fputc((color & 0b11111) * 8, outfile);
		fputc((color >> 10) * 8, outfile);
	}
	fclose(infile);
	fclose(outfile);

	return 0;
}

File: tools/src/bin/palconv.rs
use std::env::args;
use std::fs::read;
use std::fs::write;
use std::process::exit;

fn main() {
	let args: Vec<String> = args().collect();
	if args.len() != 3 {
		eprintln!("Expected two arguments");
		exit(1);
	}

	let input = read(&args[2]).unwrap_or_else(|error| {
		eprintln!("Failed to open {}: {}", args[2], error);
		exit(1);
	});

	let mut output = Vec::<u8>::new();

	for i in (0..input.len()).step_by(2) {
		let color = u16::from_le_bytes([input[i], input[i + 1]]);
		output.push(((color >> 5 & 0b11111) * 8) as u8);
		output.push(((color & 0b11111) * 8) as u8);
		output.push(((color >> 10) * 8) as u8);
	}

	write(&args[1], output).unwrap_or_else(|error| {
		eprintln!("Failed to write to {}: {}", args[1], error);
		exit(1);
	});
}